### DISCUSSION TITLE
Problem using SSO with Google

### DISCUSSION THREAD

**Discussion by stevenrenaud** [kimai/kimai#4300] [General] [answered]
### Describe the issue

My kimai instance is a fresh install 2.0.33 on ubuntu 22.04 - php 8.1 - apache 2.4 - MariaDB 10.6

It works like a charm except i cannot setup the sso with google.

I get this generic error that is not log on any log file

![image](https://github.com/kimai/kimai/assets/50968243/545c8dae-ddea-40ea-8a81-aef5c93366a6)

I follow the documentation to setup my local.yaml file

my local.yaml
------------------
```
kimai:
    saml:
        provider: google
        activate: true
        title: Login with Google
        mapping:
            - { saml: $Email, kimai: email }
            - { saml: $FirstName $LastName, kimai: alias }
            - { saml: $Title, kimai: title }
            - { saml: $AccountNumber, kimai: accountNumber }
        roles:
            resetOnLogin: true
            attribute: Groups
            mapping:
                # Insert your role-mapping here (ROLE_USER is added automatically)
                - { saml: KIMAI SYS ADMIN, kimai: ROLE_SUPER_ADMIN }
                - { saml: KIMAI ADMIN, kimai: ROLE_ADMIN }
                - { saml: KIMAI TEAMLEAD, kimai: ROLE_TEAMLEAD }
        connection:
            idp:
                entityId: 'https://accounts.google.com/o/saml2?idpid=XXXXXX'
                singleSignOnService:
                    url: 'https://accounts.google.com/o/saml2/idp?idpid=XXXXXX'
                x509cert: ' **** SINGLE OR MULTI LINE ****'
            # Your Kimai: replace https://www.example.com with your base URL
            sp:
                entityId: 'https://kimai.example.com/auth/saml/metadata'
                assertionConsumerService:
                    url: 'https://kimai.example.com/auth/saml/acs'
                #privateKey: ''
            # only set baseurl, if auto-detection doesn't work
            baseurl: 'https://kimai.example.com/auth/saml/'  #indeed it does not work !!!
            strict: true
            debug: true
            security:
                nameIdEncrypted: false
                authnRequestsSigned: false
                logoutRequestSigned: false
                logoutResponseSigned: false
                wantMessagesSigned: false
                wantAssertionsSigned: false
                wantNameIdEncrypted: false
                requestedAuthnContext: true
                signMetadata: false
                wantXMLValidation: true
                signatureAlgorithm: 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256'
                digestAlgorithm: 'http://www.w3.org/2001/04/xmlenc#sha256'
            contactPerson:
                technical:
                    givenName: 'Kimai Admin'
                    emailAddress: 'it@example.com'
                support:
                    givenName: 'Kimai Support'
                    emailAddress: 'it@example.com'
            organization:
                en:
                    name: 'Example'
                    displayname: 'Example'
                    url: 'https://kimai.example.com'
```

Apache conf
-----------------

```
<VirtualHost *:80>
    ServerName kimai
    ServerAlias kimai.example.com

    DocumentRoot /var/www/kimai/public
    <Directory /var/www/kimai/public>
        AllowOverride All

        # If you see an error like the following in your logs:
        # AH01630: client denied by server configuration: /var/www/kimai/public/
        # then you might have to exchange the "Order/Allow" rules with "Require" (see below)
        # More infos at https://httpd.apache.org/docs/2.4/de/upgrading.html

        # Apache 2.2
        #Order Allow,Deny
        #Allow from All
        
        # Apache 2.4
        Require all granted

        FallbackResource /index.php
    </Directory>

    <Directory /var/www/kimai>
        Options FollowSymlinks
    </Directory>

    # optionally disable the fallback resource for the asset directories
    # which will allow Apache to return a 404 error when files are
    # not found instead of passing the request to Symfony
    <Directory /var/www/kimai/public/bundles>
        FallbackResource disabled
    </Directory>
   
    <FilesMatch \.php$>
        SetHandler "proxy:unix:/run/php/php8.1-fpm.sock|fcgi://localhost/"
    </FilesMatch>
 
    ErrorLog /var/log/kimai_error.log
    CustomLog /var/log/kimai_access.log combined

    # optionally set the value of the environment variables used in the application
    #SetEnv APP_ENV prod
    #SetEnv APP_SECRET <app-secret-id>
    #SetEnv DATABASE_URL "mysql://db_user:db_pass@host:3306/db_name?charset=utf8mb4&serverVersion=5.7"
</VirtualHost>
```

### I already tried

- [X] I've read and searched [the documentation](https://www.kimai.org/documentation/).
- [X] I've searched for similar issues in this repository.
- [X] I've searched for similar issues in [the discussions](https://github.com/kimai/kimai/discussions).

### Kimai version

2.0.26

### How do you run Kimai?

Virtual Server or alike

### Which PHP version are you using?

8.1

### Logfile

No relevant logs

### Screenshots

Google Config

![image](https://github.com/kimai/kimai/assets/50968243/6179722e-fbc8-4dbb-b16e-b38de6aacd50)

![image](https://github.com/kimai/kimai/assets/50968243/67e1f07f-3e59-4c9a-9d57-5d3e232d3b4d)

SAML 

![image](https://github.com/kimai/kimai/assets/50968243/09676b61-8c49-40ab-96e7-8762129700ce)

**Comment by kevinpapst** [comment]
- remove `baseurl`
- use single-line for `x509cert`
- set `strict` to false.
- your config points to https, but your vhost to port 80

**Reply by stevenrenaud** [reply]
Thank you for you reply.

Try to remove base url and set strict to false with no success.
x509cert is already single line

There is a traefik redirecting https to http in between (could the issue be there ?)

**Reply by kevinpapst** [reply]
> There is a traefik redirecting https to http in between (could the issue be there ?)

Probably, if your config does not match the reality...
Try to remove Traefik, see if it works. 
The Proxy headers should be sufficient (if they are present), but I am not sure, not using such a setup myself.

**Reply by stevenrenaud** [reply]
I remount my older kimai (v1.26) and adjust it in exact same configuration (Google SAML / Traefik). There is no issue, it worked at my first attempt.

I cross check everything more than 3 times.
At this point i have no more clue on what could happen.

Any idea how i could debug this error ?

**Comment by stevenrenaud** [comment]
Hi,

I made 2 standard kimai installation  with exact same configuration for traefik, local.yaml and .env (except provider config in local.yaml that does not exists in v1) 

- One in kimai v1.30.11 (last 1.x)                 --> kimai-test.example.com
- One in kimai v2.0.34 (the one from doc)   --> kimai.example.com

In the two scenario : kimai is configure in http (80) with nginx (exact configuration from doc) serve by a traefik reverse proxy for ssl and reverse proxy purpose.

kimai-test (1.30.11) is working 
kimai        (2.0.34) is not working with the same issue as declared in post

I did an upgrade of  kimai-test (1.30.11) according to the doc to v2.0.34 and then instance broken with exact same issue

Is there any configuration that change for google sso between v1 and v2 ?

Thank you

**Reply by stevenrenaud** [reply]
I just checked every version.
The problem occurs from 2.0.32  (2.0.31 is fine)

I see this change:  Make sure that accountNumber on user has max length of 30 (for SAML registrations)
Maybe related to the fact that Google does not send accountNumber in my cas

**Reply by kevinpapst** [reply]
Can you remove these lines
https://github.com/kimai/kimai/compare/2.0.31...2.0.32#diff-8f4d7ec1b98e0ef063a86a96af29b4665d6823db1745a9ed97ed8a76109da749R137-R140
then reload bytecode cache (e.g. restart PHP) and try again.

**Reply by stevenrenaud** [reply]
Culprit is here

![image](https://github.com/kimai/kimai/assets/50968243/a8593601-f080-47d5-b08d-ccf93668b372)

**Reply by stevenrenaud** [reply]
i suppose something like this would be more appropriate

``
                if (!isset($attributes[$key])) {
                    throw new \RuntimeException('Missing user attribute value: ' . $key);
                }
``
But,
  these runtime exception are very hard to debug (no log, no browser console, no error on network, no error on webserver)
  why block the loggin process, if a mapped value is null Name, Employee ID, ... might be null

**Reply by kevinpapst** [reply]
Can you replace the code with and check if it works:

                if (\array_key_exists(0, $attributes[$key])) {
                    $results[] = $attributes[$key][0];
                }

**Reply by stevenrenaud** [reply]
Yes it works.

**Reply by kevinpapst** [reply, marked as answer]
Not sure which code is the best to fix that issue.
I did it this way now: https://github.com/kimai/kimai/pull/4302/commits/868ee063c6640dee230ffe1379378701fdb70e93
Will release tomorrow.

Thanks for being persistent and your testing input, super helpful 👍

**Reply by stevenrenaud** [reply]
No problem.
Thank you for your work and time in Kimai. Fantastic tool !

I might be interested to participate in connection with tools like ERP, ...
Maybe some make plugin [make.com](https://make.com)
Let me know if there migth be somme interrests

**Reply by kevinpapst** [reply]
An integration with Zappier or (if that is a similar tool) "Make" would be awesome 👍
